Add theme properties for candidate label/text/comment color

DEAN-Cherry opened the pull request:
Adding new `Theme` propety `candidateTextColor` to control candidate text color. The default value is as same as the `keyTextColor`

---
添加新的 `Theme` 属性 `candidateTextColor`，以控制候选文本颜色。默认值与 `keyTextColor` 相同。

rocka reviewed (changes requested):
Thanks for your contribution! However I can't merge this PR in current state since it does not meet some of the design pattern in the code base. I'll explain all of my considerations.

To begin with, `Theme` is not an arbitary JSON object, it's a versioned structure with migration strategies:

https://github.com/fcitx5-android/fcitx5-android/blob/78cfe81769f673e9c9ec8e1e3557342106843fd4/app/src/main/java/org/fcitx/fcitx5/android/data/theme/CustomThemeSerializer.kt#L55-L79

which means that you should bump the theme version and implement migration strategy when adding new properties, in order to keep compatibility with older theme files.

Additionally, I feel like it's kind of not "worth it" to bump the version for just one new property. In fact I'm also planning to add more theme properties in this version, including `toolbarIconColor`, `candidateLabelColor`, `candidateWindowBorderColor` and so on.

Seeing you've also submitted https://github.com/fcitx5-android/theme-designer/pull/5 , I suggest following changes:

1. Add 2 additional theme properties `candidateLabelColor` and `candidateCommentColor`
2. Bump theme version to `2.1`
3. Apply those new theme properites to CandidateView as well
    https://github.com/fcitx5-android/fcitx5-android/blob/78cfe81769f673e9c9ec8e1e3557342106843fd4/app/src/main/java/org/fcitx/fcitx5/android/input/candidates/floating/LabeledCandidateItemUi.kt#L29-L42

DEAN-Cherry commented:
Thank you so much for the detailed review! I have incorporated all your suggestions and modified the code. Could you please take another look? Again, thanks a lot for your thorough review!

rocka commented on app/src/main/java/org/fcitx/fcitx5/android/input/candidates/floating/LabeledCandidateItemUi.kt:
```diff
@@ -27,8 +27,8 @@ class LabeledCandidateItemUi(
     fun update(candidate: FcitxEvent.Candidate, active: Boolean) {
-        val fg = if (active) theme.genericActiveForegroundColor else theme.keyTextColor
-        val altFg = if (active) theme.genericActiveForegroundColor else theme.altKeyTextColor
+        val fg = if (active) theme.genericActiveForegroundColor else theme.candidateLabelColor
```
Just use `candidateLabelColor` and `candidateTextColor` on "label" and "text" respectively

```suggestion
        val labelFg = if (active) theme.genericActiveForegroundColor else theme.candidateLabelColor
        val fg = if (active) theme.genericActiveForegroundColor else theme.candidateTextColor
```

rocka commented on app/src/main/java/org/fcitx/fcitx5/android/input/candidates/floating/LabeledCandidateItemUi.kt:
```diff
@@ -27,8 +27,8 @@ class LabeledCandidateItemUi(
+        val fg = if (active) theme.genericActiveForegroundColor else theme.candidateLabelColor
+        val altFg = if (active) theme.genericActiveForegroundColor else theme.candidateCommentColor
         root.text = buildSpannedString {
             color(fg) { append(candidate.label) }
```
```suggestion
            color(labelFg) { append(candidate.label) }
```

rocka commented on app/src/main/java/org/fcitx/fcitx5/android/data/theme/CustomThemeSerializer.kt:
```diff
@@ -80,8 +99,8 @@ object CustomThemeSerializer : JsonTransformingSerializer<Theme.Custom>(Theme.Cu
-    private const val CURRENT_VERSION = "2.0"
-    private const val FALLBACK_VERSION = "1.0"
+    private const val CURRENT_VERSION = "2.1"
+    private const val FALLBACK_VERSION = "2.0"
```
```suggestion
    private const val FALLBACK_VERSION = "1.0"
```

`FALLBACK_VERSION` should not be changed, it's the "fallback" version when importing themes without "version" key

rocka commented on app/src/main/java/org/fcitx/fcitx5/android/data/theme/CustomThemeSerializer.kt:
```diff
@@ -55,6 +55,25 @@ object CustomThemeSerializer : JsonTransformingSerializer<Theme.Custom>(Theme.Cu
         listOf(
+            MigrationStrategy("2.1") {
+                JsonObject(it.toMutableMap().apply {
+                    if (get("candidateTextColor") != null) {
+                        put("candidateTextColor", getValue("candidateTextColor"))
+                    } else {
+                        put("candidateTextColor", getValue("keyTextColor"))
+                    }
+                    if (get("candidateLabelColor") != null) {
+                        put("candidateLabelColor", getValue("candidateLabelColor"))
+                    } else {
+                        put("candidateLabelColor", getValue("keyTextColor"))
+                    }
+                    if (get("candidateCommentColor") != null) {
+                        put("candidateCommentColor", getValue("candidateCommentColor"))
+                    } else {
+                        put("candidateCommentColor", getValue("altKeyTextColor"))
+                    }
```
```suggestion
                    put("candidateTextColor", getValue("keyTextColor"))
                    put("candidateLabelColor", getValue("keyTextColor"))
                    put("candidateCommentColor", getValue("altKeyTextColor"))
```

Filling all the missing theme properties from 2.0 to 2.1 should be enough. If a theme happens to have those properties, it should be versioned as 2.1

rocka reviewed (changes requested):
Some little suggestions: simplify the migration strategy and actually apply `candidateLabelColor`

DEAN-Cherry commented:
Thanks for your review, I didn't realize I could use `candidateTextColor` to specify the text color for LabelUI. I overlooked that.

![PixPin_2024-12-28_14-40-55](https://github.com/user-attachments/assets/50e57874-1e3d-468f-97f3-74f12d8e2caa)
